System.Linq.Parallel Test - initial cleanup

Clockwork-Muse opened the pull request:
Initial PR for cleaned up PLINQ tests.

note: I don't consider myself done with this by any means.  In particular, there's a number of things I want to add to the combinatorial tests (like cancellation).

All tests are passing - all 200K+ of them (most of that is the combinatorial tests).  The original version didn't appear to have many, due to not using `Theory`.  `OuterLoop` will take a _very_ long time to run, ~10 minutes on my machine(innerloop runs in about 90 seconds).

@rafidka - you should be able to grab things now.  You'll actually want [this other branch instead](https://github.com/Clockwork-Muse/corefx/tree/516_System.Linq.Parallel_FullyOrdered), which assumes more ordered output than PLINQ currently delivers (but I'm not sure if it's a bug...).

I'm going to be in Scotland for two weeks, so my ability to reply/deal with changes from the code review in that time will be limited.

Possibly interested people: 
@stephentoub , @alfredoMS , @svick, @weshaggard

Clockwork-Muse commented:
Hmm... looks like Jenkins didn't like how long it was taking the combinatorial tests (the bulk of the tests - 150K +).  Marking them outerloop 'fixes' this, something I debated from the start.

weshaggard commented:
@Clockwork-Muse thanks for helping out with the test coverage here. I've not looked at code review at all yet but just from the description I can tell you we want InnerLoop tests to run in a few seconds < 5 so 90 seconds is definitely too long, so yes I do think we should be marking most of these as OuterLoop. 

Also after the reviews are done we are going to ask you to squash the commits.

Maxwe11 commented:
@dotnet-bot test this please

Clockwork-Muse commented:
@weshaggard - I'm assuming you mean for all tests in that category, not individually.  I'll have to see what I can knock down (should be simple, just changing counts).  

I can't check this now, but I feel like the original innerloop tests were still in the 20 second range for this library (with outerloop at about 90).  

I'd imagine there's going to need to be a rebase anyways, so squashing the commits isn't unexpected.

weshaggard commented:
@Clockwork-Muse our goal is to have all InnerLoop tests for a given library to run in around 5 seconds, I know that isn't the case for all libraries right now but that is our target time.

stephentoub commented on src/System.Linq.Parallel/src/System/Linq/Parallel/Utils/ExceptionAggregator.cs:
```diff
@@ -116,13 +115,16 @@ internal static void ThrowOCEorAggregateException(Exception ex, CancellationStat
+            else if (ex is AggregateException)
+            {
+                throw ((AggregateException)ex).Flatten();
+            }
```
This is changing the product behavior.  Why?

Clockwork-Muse replied:
Because it was nesting `AggregateException`s (I can't recall what from now, but it should be one of the immediately surrounding commits).  It makes whatever that method was behave the same as everything else - only a single level of AE before the offending exception.I actually want to investigate moving use of the method in general into other places (with or without this change), since there's similar cancellation checking elsewhere.  Other than that, `QueryTaskState` is doing similar flattening (although the situation isn't completely identical).

stephentoub replied:
But this was by-design behavior, and it matches the desktop implementation, right?  We shouldn't be changing the product behavior as part of a set of test changes like this.

Clockwork-Muse replied:
I can't remember off hand, but I assume so.I didn't know if it was by-design, or by-accident.  Nothing else throws nested AEs, so I was assuming bug.  I can definitely pull it for more discussion in a later PR/issue or something.

stephentoub commented:
Thanks for the effort invested in these tests, @Clockwork-Muse.  I'll go through it, but it'll take some time; 188 is a lot of commits ;)

Clockwork-Muse commented:
@stephentoub - heh, sorry.  That's part of why I'm doing this _now_, instead of later, after more work...

Clockwork-Muse commented:
@weshaggard - managed to get innerloop down to ~8 seconds on my machine.  I don't know that I can do much better than that (and I'm twitchy as is).
@stephentoub - removed the `AggregateException` un-nesting, will probably post as new issue later.

stephentoub commented on src/System.Linq.Parallel/src/System/Linq/ParallelEnumerable.cs:
```diff
@@ -5354,18 +5351,20 @@ public static ParallelQuery<TResult> OfType<TResult>(this ParallelQuery source)
         /// </exception>
         public static ParallelQuery<TResult> Cast<TResult>(this ParallelQuery source)
         {
+            if (source == null) throw new ArgumentNullException("source");
```
While this makes sense and is arguably how it should have been in the first place, it's also a behavioral change from the desktop.  @AlfredoMS, is this ok?

Clockwork-Muse replied:
I know, and I debated about changing it after dealing with the `SequenceEqual` stuff, but this felt more obviously a bug...

Isn't the desktop stuff eventually pulling these changes in anyways?  Well, I guess "eventually" might be its own problem...

stephentoub replied:
> Isn't the desktop stuff eventually pulling these changes in anyways?

For perf improvements and the like, hopefully. But breaking changes are something that need to be discussed in detail.

I'd suggest not making this change here; it should be done as part of a separate PR if we want to make the change.

stephentoub commented on src/System.Linq.Parallel/src/System/Linq/ParallelEnumerable.cs:
```diff
@@ -5354,18 +5351,20 @@ public static ParallelQuery<TResult> OfType<TResult>(this ParallelQuery source)
-        /// <typeparam name="TResult">The type to convert the elements of <paramref name="source"/> to.</typeparam>
-        /// <param name="source">The sequence that contains the elements to be converted.</param>
+        /// <typeparam name="TResult">The stronger type to convert the elements of <paramref name="source"/> to.</typeparam>
+        /// <param name="source">The sequence that contains the weakly typed elements to be converted.</param>
```
You have "weakly-typed" hyphenated everywhere but here.

stephentoub commented on src/System.Linq.Parallel/tests/Combinatorial/CancellationParallelQueryCombinationTests.cs:
```diff
@@ -0,0 +1,230 @@
+        [Theory]
+        [MemberData("UnaryOperators")]
+        [MemberData("BinaryOperators")]
+        public static void Aggregate_OperationCancelledException(LabelledOperation source, LabelledOperation operation)
```
"Canceled" is spelled with one 'l'.

stephentoub commented on src/System.Linq.Parallel/tests/Helpers/Labelled.cs:
```diff
@@ -0,0 +1,42 @@
+
+namespace Test
+{
+    internal static class Labelled
```
"Labeled" has one "l".

Clockwork-Muse replied:
I have no idea why I like this spelling...

stephentoub commented on src/System.Linq.Parallel/tests/Combinatorial/CancellationParallelQueryCombinationTests.cs:
```diff
@@ -0,0 +1,230 @@
+        [MemberData("BinaryOperators")]
+        public static void Aggregate_OperationCancelledException(LabelledOperation source, LabelledOperation operation)
+        {
+            CancellationTokenSource cs = new CancellationTokenSource();
```
If all you want is an already canceled token, you can just use ```new CancellationToken(true)```

Clockwork-Muse replied:
Good to know, thanks!

stephentoub replied:
Although seeing your usage of it further in the tests, probably makes sense to stick with what you have.  You're verifying that the OCE's that emerge contain the same token, and that can be better verified when the CT is associated with a CTS instance as you have it.

Clockwork-Muse replied:
...Interestingly, it turns out that It still compares equal.  That feels slightly worrying, actually.
Now I'm wondering if I should go back to the old version and change it to check for reference equals(?)...

stephentoub replied:
That was part of my follow-up point.  ```new CancellationToken(true).Equals(new CancellationToken(true))```, which is desirable in general but could lead to false positives if the library did the wrong thing.  By using a new CancellationTokenSource as you were doing originally, the only way the comparison on teh exception's Token would match is if it came from the same CTS, whereas with ```new CancellationToken(true)``` as the input, the OCE could manufacture another ```new CancellationToken(true)``` and it would match.

stephentoub commented on src/System.Linq.Parallel/tests/Combinatorial/CancellationParallelQueryCombinationTests.cs:
```diff
@@ -0,0 +1,230 @@
+
+namespace Test
+{
+    [Trait("category", "outerloop")]
```
It doesn't work to just use ```[OuterLoop]```?  If not, @krwq, seems like we should fix that, right?

Clockwork-Muse replied:
First thing I tried!
`"Attribute 'OuterLoop' is not valid on this declaration type.  It is only valid on 'method' declarations"`.
Debated about reporting/fixing it myself, but wasn't sure it was a deliberate choice or not.

stephentoub commented on src/System.Linq.Parallel/tests/Combinatorial/SourcesAndOperators.cs:
```diff
@@ -0,0 +1,304 @@
+{
+    public partial class ParallelQueryCombinationTests
+    {
+        private const int defaultStart = 8;
```
Per corefx design guidelines, constants should be PascalCased.

Clockwork-Muse replied:
You're lucky I didn't FOLLOW_JAVA_STANDARDS.

stephentoub commented on src/System.Linq.Parallel/tests/Combinatorial/ParallelQueryCombinationTests.cs:
```diff
@@ -0,0 +1,1523 @@
+        [MemberData("BinaryOperators")]
+        public static void Aggregate_Result(LabelledOperation source, LabelledOperation operation)
+        {
+            Assert.Equal(Functions.SumRange(defaultStart, defaultSize) + 17,
```
What's the relevance of "17" here?  Is it just an arbitrary number?  I'd put it in a named constant to better describe it, and because it's used multiple times and presumably has to match each time.

Clockwork-Muse replied:
Arbitrary.  I've moved it to the stand-alone Aggregate tests, as it fits there better.

stephentoub commented on src/System.Linq.Parallel/tests/Combinatorial/ParallelQueryCombinationTests.cs:
```diff
@@ -0,0 +1,1523 @@
+        {
+            AggregateException ae = null;
+            // Concat seems able to return the first element when the left query does not fail ("first" query).
+            // This test might be flaky in the case that it decides to run the right query too...
```
It would be a breaking change to start enumerating an enumerable that wasn't previously getting enumerated.  You say the test might be flaky in that case... can we make it so that it definitively breaks if that were to happen?

Clockwork-Muse replied:
Sorry, what I was referring to was the fact that, per discussion after #584 was closed, I don't necessarily have a guarantee that `Concat` always calls the left query first (or before the right might start spooling), when unordered.  Or at least that's the way I read it at the time.

stephentoub commented on src/System.Linq.Parallel/tests/Helpers/IntegerRangeSet.cs:
```diff
@@ -0,0 +1,125 @@
+        {
+            Assert.InRange(entry, _start, _start + _seen.Length - 1);
+
+            return _seen[entry - _start] ^= true;
```
Should this instead throw if _seen[entry-_start] is already true?  I understand your goal with the xor, but if the same value were to be added three times, for example, it would end up as true even though we'd erroneously seen the same value multiple times, right?

Clockwork-Muse replied:
I was trying to be clever and one-line it, even though I knew this was a potential problem.  Since 99% of the time I was checking the immediate return value anyways (would throw on the second value) I didn't consider it much of a risk...

stephentoub replied:
Ok, though for code used so frequently across such a large number of tests, seems like a fine place to be less clever and pay the extra line for more robust testing.

Clockwork-Muse replied:
Yeah, probably.
It's turning out that BitArray of course doesn't synchronize its operations, and of course many (all?) of my tests are operating in threaded contexts... and because it's using `byte`s to store the actual data things are going sideways.  I'll have to decide if locking is worth the space savings.

stephentoub commented on src/System.Linq.Parallel/tests/Helpers/IntegerRangeSet.cs:
```diff
@@ -0,0 +1,125 @@
+    // Dummy psuedo-set for verifying we've seen all of a range of integers, and only once.
+    internal class IntegerRangeSet : IEnumerable<KeyValuePair<int, bool>>
+    {
+        private bool[] _seen;
```
This is ok, but you could also use a BitArray.

Clockwork-Muse replied:
Hmmm....
I originally wrote this to be "self sufficient", and not rely on other libraries for helpers (which is why I didn't use a set).  Although I guess backing it with a BitArray would save a lot of space (8x+ or something like that)...

stephentoub replied:
Noble goal, but I think it'd be better to use the core collection types rather than writing equivalent custom collection types.

Clockwork-Muse replied:
Yeah, probably.
 It's turning out that BitArray of course doesn't synchronize its operations, and of course many (all?) of my tests are operating in threaded contexts... and because it's using  byte s to store the actual data things are going sideways. I'll have to decide if locking is worth the space savings.

stephentoub replied:
FWIW, this implementation isn't thread-safe either (I just commented on that in the ForAll tests), as multiple calls could be made concurrently for the same value, both of which could transition a false to true, even though one should fail.

stephentoub commented on src/System.Linq.Parallel/tests/Helpers/IntegerRangeSet.cs:
```diff
@@ -0,0 +1,125 @@
+            return _seen[entry - _start] ^= true;
+        }
+
+        public bool Complete()
```
Maybe "IsCompete"?  As it's currently named it sounds like it's mutating something.

stephentoub commented on src/System.Linq.Parallel/tests/Combinatorial/ParallelQueryCombinationTests.cs:
```diff
@@ -0,0 +1,1523 @@
+        [MemberData("BinaryOperators")]
+        public static void GroupBy_NotPipelined(LabelledOperation source, LabelledOperation operation)
+        {
+            int seenKey = defaultStart / 8;
```
It'd be good to put 8 into a named constant in all of these places.

stephentoub commented on src/System.Linq.Parallel/tests/Combinatorial/ParallelQueryCombinationTests.cs:
```diff
@@ -0,0 +1,1523 @@
+                    Assert.True(seen.Add(p.Key / 2));
+                    Assert.Equal(p.Key, p.Value * 2);
+                });
+            Assert.True(seen.Complete());
```
In all of the places you do ```Assert.True(seen.Complete())```, did you consider instead just having a call like ```seen.AssertComplete()```, where the implementation would include in the assert message information about what was missing (or even duplicated if you chose to change the implementation of the collection)?

stephentoub commented on src/System.Linq.Parallel/tests/Combinatorial/SourcesAndOperators.cs:
```diff
@@ -0,0 +1,304 @@
+
+        private static IEnumerable<LabelledOperation> UnorderedRangeSources()
+        {
+            // The difference between this and the existing sources is I need more control over the range creation.
```
Nit: can we avoid the use of "I" in the code?  It'd be better to use "we" or simply avoid the use of pronouns.

stephentoub commented on src/System.Linq.Parallel/tests/Combinatorial/SourcesAndOperators.cs:
```diff
@@ -0,0 +1,304 @@
+                    yield return new object[] { source, operation };
+                }
+
+                // For Concat, since either one can be ordered the other side has to be tested
```
This is commented out because... of one of the issues you opened?  If so, can you add the issue #?

Clockwork-Muse replied:
I'm still waiting to hear if this is an actual bug or not, but yeah, I can add it.

stephentoub commented on src/System.Linq.Parallel/tests/DegreeOfParallelismTests.cs:
```diff
@@ -1,140 +1,86 @@
-            RunDopPipeliningTestCore(1000000, 512);
+            //int minWorker, minAsyncIO;
+            //ThreadPool.GetMinThreads(out minWorker, out minAsyncIO);
+            //ThreadPool.SetMinThreads(degree, minAsyncIO);
```
Why is this commented out?  You should consider reinstating it, along with code to reset the minimums back to what they were if you changed them due to them being too low.  If you don't, it could take at least seconds for the thread pool to otherwise ramp up to 32 threads, and it could take minutes for it to ramp up to 512 threads, given that it'd likely only be doing so based on its starvation policy which will only inject 1 or 2 threads per second if it sees things stalled.

Clockwork-Muse replied:
When I last wrote it, `ThreadPool` (or maybe just those methods) wasn't available.  I'll check again.

stephentoub replied:
Hmm, if it's not exposed in the contract, @ericeil / @weshaggard, what's the recommended way for folks to handle situations like this, where concurrent work really needs a particular minimum number of threads to be available in the pool?

ericeil replied:
We had to remove these because they are not portable to other thread pool implementations; not every thread pool gives this kind of control.  I believe the issue here is limited to test scenarios only; production code should not be making assumptions about thread-injection policy.  Given that, I could see value in adding some sort of internal contract to control the number of threads in the thread pool, which we only support on some platforms, and perhaps the test could fall back to some other behavior on platforms that don’t support this.  However, this is certainly not something we support now.

Eric

From: Stephen Toub [mailto:notifications@github.com]
Sent: Tuesday, June 2, 2015 7:46 AM
To: dotnet/corefx
Cc: Eric Eilebrecht
Subject: Re: [corefx] System.Linq.Parallel Test - initial cleanup (#1599)


In src/System.Linq.Parallel/tests/DegreeOfParallelismTests.cs<https://github.com/dotnet/corefx/pull/1599#discussion_r31530064>:

>          {

> -            RunDopPipeliningTestCore(1000000, 1);

> -            RunDopPipeliningTestCore(1000000, 4);

> -            RunDopPipeliningTestCore(1000000, 64);

> -            RunDopPipeliningTestCore(1000000, 512);

> +            //int minWorker, minAsyncIO;

> +            //ThreadPool.GetMinThreads(out minWorker, out minAsyncIO);

> +            //ThreadPool.SetMinThreads(degree, minAsyncIO);

Hmm, if it's not exposed in the contract, @ericeil<https://github.com/ericeil> / @weshaggard<https://github.com/weshaggard>, what's the recommended way for folks to handle situations like this, where concurrent work really needs a particular minimum number of threads to be available in the pool?

—
Reply to this email directly or view it on GitHub<https://github.com/dotnet/corefx/pull/1599/files#r31530064>.

stephentoub commented on src/System.Linq.Parallel/tests/DegreeOfParallelismTests.cs:
```diff
@@ -1,140 +1,86 @@
+        [MemberData("DegreeData", new[] { 1, 4 }, new object[] { })]
+        [MemberData("DegreeData", new[] { 32 }, new[] { 4 })]
+        // With more items and threads, the total runtime can far exceed the total delay.
+        // This does not appear to be solely due to thrashing - thread creation appears overly expensive.
```
PLINQ doesn't create new threads, it queues tasks to run on the thread pool.  But the pool is relatively conservative about injecting new threads, so if lots of tasks are queued and all of those tasks need to run concurrently in order to make forward progress, the thread pool may only slowly ramp up to the required number of threads.

stephentoub commented on src/System.Linq.Parallel/tests/ParallelEnumerableTests.cs:
```diff
@@ -0,0 +1,240 @@
+            yield return new object[] { int.MaxValue, 0 };
+            yield return new object[] { int.MaxValue, 1 };
+            yield return new object[] { int.MaxValue - 8, 8 + 1 };
+        }
```
Do you want to test the other end of the spectrum as well, e.g. ```yield return new object[] { int.MinValue, 1 }```?

stephentoub commented on src/System.Linq.Parallel/tests/PlinqDelegateExceptions.cs:
```diff
@@ -77,9 +77,9 @@ public static void OrderBy()
         /// <summary>
-        /// Heavily exercises OrderBy in the face of user-delegate exceptions. 
+        /// Heavily exercises OrderBy in the face of user-delegate exceptions.
         /// On CTP-M1, this would deadlock for DOP=7,9,11,... on 4-core, but works for DOP=1..6 and 8,10,12, ...
```
We can remove comments about CTPs; those were all pre-.NET 4.

Clockwork-Muse replied:
From the look of it, I haven't done any actual work on this file (just run cleanup over it in batch).  I'm probably going to revert this file for now, in favor of dealing with it more later.

stephentoub replied:
Ok, good idea.  Probably best to do that in general.

stephentoub commented on src/System.Linq.Parallel/tests/QueryOperators/AggregateTests.cs:
```diff
@@ -0,0 +1,199 @@
+        }
+
+        [Fact]
+        public static void Aggregate_OperationCancelledException()
```
How does this test differ in purpose from the equivalent one in the combinatorial tests?  Won't the combinatorial one be run on similar sources (and more)?

Clockwork-Muse replied:
The combinatorial tests don't check for `InvalidOperationException` (mostly because it doesn't return an empty source) or `OperationCanceledException` (yet - stuff I'm writing after I get this first batch done, and even then it's throwing one level down).

stephentoub replied:
Ok.

stephentoub commented on src/System.Linq.Parallel/tests/QueryOperators/AggregateTests.cs:
```diff
@@ -0,0 +1,199 @@
+        {
+            ParallelQuery<int> query = labelled.Item;
+            int actual = query.Aggregate(
+                0,
```
I may be misremembering, but I think that all of the aggregate tests I've seen thus far have used 0 as a seed.  Are there any tests for other values? e.g. an accumulation that does multiplication with a seed of 1? an accumulation that accumulates into a ```List<T>``` and returns a ```List<T>``` that combines all the partial ```List<T>``` instances? etc.

Clockwork-Muse replied:
No, there aren't.  Something I probably should have done to start with.

stephentoub commented on src/System.Linq.Parallel/tests/QueryOperators/AggregateTests.cs:
```diff
@@ -0,0 +1,199 @@
+            Assert.Throws<ArgumentNullException>(() => ParallelEnumerable.Range(0, 1).Aggregate<int, int, int>(() => 0, (i, j) => i, (i, j) => i, null));
+        }
+    }
+}
```
What about tests that verify:
- Aggregate can use multiple threads concurrently
- The delegates are invoked an appropriate number of times based on the number of threads / elements used
- etc.
Or would those be more appropriate elsewhere, e.g. in the DOP tests?  Just seemed relevant here, as Aggregate has logic to combine partial aggregations into the total aggregation, and we want to make sure that's appropriately exercised when the logic to combine the partials is non-trivial, e.g. building a new list from partial lists.

stephentoub commented on src/System.Linq.Parallel/tests/QueryOperators/AllTests.cs:
```diff
@@ -0,0 +1,113 @@
+{
+    public class AllTests
+    {
+        public static IEnumerable<object[]> OnlyOneData(object[] counts)
```
I'm not understanding the name of this function; what is the purpose of this?

Clockwork-Muse replied:
Generates data for tests where a differing item (ie, `false` for a collection of `true` items) occurs at the start, middle, and end of the collection..

I probably should comment some of these more (at all?!)

stephentoub commented on src/System.Linq.Parallel/tests/QueryOperators/AllTests.cs:
```diff
@@ -0,0 +1,113 @@
+        [Fact]
+        public static void All_OperationCancelledException()
+        {
+            CancellationTokenSource cs = new CancellationTokenSource();
```
Same question about the duplication with the combinatorial tests... I'll stop asking about these.  If they serve a different purpose, ok.  But if they're actually duplicative, we should probably just delete these.

Clockwork-Muse replied:
Hmm... for the combinatorial, I generally did only one of the 'outer' functions each, just as a SAN check things worked in combination.  I don't test only one item being true in the combinatorial (just all/none).

stephentoub replied:
Ok.

stephentoub commented on src/System.Linq.Parallel/tests/QueryOperators/AnyTests.cs:
```diff
@@ -0,0 +1,150 @@
+        {
+            Assert.True(InfiniteEnumerable().AsParallel().Any());
+            Assert.True(InfiniteEnumerable().AsParallel().Any(x => true));
+        }
```
I might have missed it, but what about Any on an empty enumerable? Same question for All.

Clockwork-Muse replied:
You missed it.  In fact, I test (almost?) everything with an empty source (sometimes this means they get `Assert.Throws...` instead)

stephentoub replied:
Ok :smile:

stephentoub commented on src/System.Linq.Parallel/tests/QueryOperators/ForAllTests.cs:
```diff
@@ -0,0 +1,59 @@
+        {
+            ParallelQuery<int> query = labelled.Item;
+            IntegerRangeSet seen = new IntegerRangeSet(0, count);
+            query.ForAll<int>(x => Assert.True(seen.Add(x)));
```
The logic in IntegerRangeSet isn't thread-safe, so it's possible this isn't doing the test you're hoping for, e.g. two iterations could both invoke seen.Add concurrently with the same value and both succeed.

Clockwork-Muse replied:
I'm aware it's not threadsafe.  I initially decided that the library was probably well-behaved enough it probably wouldn't be an issue, or that any duplicate entries probably had a good enough chance of mutating something already set.
Although since it's not `volatile` Booleans... hmmmm... 
I was really hoping to avoid the overhead of locking.  Especially because that's going to increase test run time.

Clockwork-Muse replied:
Done, and nowhere near what I feared, if it even takes any extra time.

Clockwork-Muse replied:
@stephentoub  - nope nope nope, locking a BitArray to do this safely dies in the face of outerloop tests.  I think I'm going to need my own version that locks by partition or something.  Using Interlocked.Increment for the counting set also runs into problems at scale (although I use that class less).

stephentoub replied:
> dies in the face of outerloop tests

I just grabbed your branch and tried it both as-is for inner loop and outer loop and then with a slight variant of your IntegerRangeSet:
```C#
internal class IntegerRangeSet : IEnumerable<KeyValuePair<int, bool>>
{
    private int[] _seen;
    private int _start;

    public IntegerRangeSet(int start, int count)
    {
        _start = start;
        _seen = new int[count];
    }

    public bool Add(int entry)
    {
        Assert.InRange(entry, _start, _start + _seen.Length - 1);
        return Interlocked.CompareExchange(ref _seen[entry - _start], 1, 0) == 0;
    }

    public bool Complete()
    {
        return _seen.All(x => x == 1);
    }
    ...
```
Prior to using Interlocked.CompareExchange, inner loop took 5 seconds and outer loop took 491 seconds.  After this switch, inner loop took 7 seconds and outer loop took 498.  That's only a 1.5% increase in outer loop time... you're seeing differently?  I could imagine using a lock with BitArray could be significantly more, but my comment about BitArray was just in passing, and before I realized you had concurrency problems.

Clockwork-Muse replied:
Sorry, it turns out I had a bit of a Shemiel (sp?) The Painter algorithm in the returned enumerator.  I'm seeing similar results on my machine for the interlocked version (although mine takes twice as long due to hardware).  Now re-investigating BitArray.

stephentoub replied:
("Schlemiel" or "Shlemiel"... it's a Yiddush word and so can multiple spellings when transliterated to English.)

Clockwork-Muse replied:
Okay, looks like locking (ReaderWriterLockSlim) is about the same, +1%.
Moral of the story: IEnumerable.ElementAt(...) isn't guaranteed to work in O(n) time over BitArrays (probably because LINQ doesn't have special handling for it).  Of course, it doesn't help it's not exposing AsQueryable at the moment either...

stephentoub commented on src/System.Linq.Parallel/tests/QueryOperators/GroupByTests.cs:
```diff
@@ -0,0 +1,458 @@
+        public static void GroupBy_Unordered(Labelled<ParallelQuery<int>> labelled, int count)
+        {
+            ParallelQuery<int> query = labelled.Item;
+            IntegerRangeSet groupsSeen = new IntegerRangeSet(0, Math.Min(count, 8));
```
As noted elsewhere, it'd be good to move constants like this 8 into a named constant.

stephentoub commented on src/System.Linq.Parallel/tests/QueryOperators/MaxTests.cs:
```diff
@@ -0,0 +1,416 @@
+            oe = Assert.Throws<OperationCanceledException>(() => ParallelEnumerable.Repeat((decimal?)0, 2).WithCancellation(cs.Token).Max());
+            Assert.Equal(cs.Token, oe.CancellationToken);
+            oe = Assert.Throws<OperationCanceledException>(() => ParallelEnumerable.Repeat((decimal?)0, 2).WithCancellation(cs.Token).Max(x => x));
+            Assert.Equal(cs.Token, oe.CancellationToken);
```
This pattern occurs a lot throughout these tests.  What about adding a helper like:
```C#
internal static void AssertOperationCanceledException(Action action, CancellationToken cancellationToken)
{
    var oce = Assert.Throws<OperationCanceledException>(action);
    Assert.Equals(cancellationToken, oce.CancellationToken);
}
```
such that all of these two-line call sites come down to a single line?

stephentoub commented on src/System.Linq.Parallel/tests/QueryOperators/MaxTests.cs:
```diff
@@ -0,0 +1,416 @@
+        public static void Max_AggregateException()
+        {
+            AggregateException ae = Assert.Throws<AggregateException>(() => ParallelEnumerable.Repeat(0, 1).Select((Func<int, int>)(x => { throw new DeliberateTestException(); })).Max());
+            Assert.All(ae.InnerExceptions, e => Assert.IsType<DeliberateTestException>(e));
```
A similar helper could be done for AssertAggregateException.

stephentoub commented on src/System.Linq.Parallel/tests/QueryOperators/MaxTests.cs:
```diff
@@ -0,0 +1,416 @@
+
+            // This fails if the maximum's most significant digit is (the text representation starts with) a '9'
+            int digits = 0;
+            for (int m = max; m > 0; m = m / 10)
```
Are you just computing ```Math.Log(max, 10)```?

Clockwork-Muse replied:
...yes, yes I am.  It's been too long since high school.

stephentoub commented on src/System.Linq.Parallel/tests/QueryOperators/OrderByThenByTests.cs:
```diff
@@ -0,0 +1,949 @@
+
+        private static int[] GetRandomInput(int count)
+        {
+            Random source = new Random();
```
Should we use a fixed seed to make things more reproducable?

stephentoub commented on src/System.Linq.Parallel/tests/QueryOperators/SelectSelectManyTests.cs:
```diff
@@ -0,0 +1,641 @@
+        {
+            yield return Labelled.Label("Array", (Func<int, int, IEnumerable<int>>)((start, count) => Enumerable.Range(start * count, count).ToArray()));
+            yield return Labelled.Label("Enumerable.Range", (Func<int, int, IEnumerable<int>>)((start, count) => Enumerable.Range(start * count, count)));
+            yield return Labelled.Label("ParallelEnumerable.Range", (Func<int, int, IEnumerable<int>>)((start, count) => ParallelEnumerable.Range(start * count, count)));
```
Should we add some types besides arrays, enumerables, and parallel enumerables? e.g. add a list, a read-only collection, etc.?

stephentoub commented on src/System.Linq.Parallel/tests/System.Linq.Parallel.Tests.csproj:
```diff
@@ -24,38 +24,68 @@
-    <Compile Include="CastOfTypeTests.cs" />
-    <Compile Include="ContainsTests.cs" />
-    <Compile Include="CountTests.cs" />
+    <Compile Include="Helpers\Comparers.cs" />
```
Could you sort this list of files in alphabetical order?

Clockwork-Muse replied:
Sure - I'm assuming that means the full path.

stephentoub replied:
Yes, please.

stephentoub commented on src/System.Linq.Parallel/tests/WithCancellationTests.cs:
```diff
@@ -185,17 +185,15 @@ public static void CTT_NonSorting_SynchronousMergerEnumeratorDispose()
-
         /// <summary>
-        /// 
+        ///
```
Are all of the changes to this file just formatting, i.e. removing a space here and there?  I think you alluded to this separately, but we should probably keep all such trivial formatting changes as part of a separate cleanup done by a tool.

Clockwork-Muse replied:
Yeah, I'll probably revert this file too, because I don't think I've touched it otherwise.

stephentoub commented:
@Clockwork-Muse, thanks for this huge amount of work.  I took a pass through it and left a bunch of comments, but on the whole it looks good to me.  I'll review your subsequent changes to; just let me know when you've submitted them.  And as discussed earlier, when we're ready to go with this, we should squash this down.

@AlfredoMS, do you want to take a pass through as well?

Clockwork-Muse commented:
@stephentoub - tests (MaxTests.Max_Other and MinTests.Min_Other) are currently failing on Linux due to the hyphen (`-`).  Given an ASCII chart table, it looks like I've written the test incorrectly (I was expecting it to be greater than a 9, instead it's less than a 0), however Windows seems to be running it fine as-is.  I haven't been able to get it to fail on my local win8 machine (in prep for fixing it, make sure it fails the same way), and the file does seem to be encoded in UTF-8 (with signature).  Any recommendations?

stephentoub commented:
You probably saw, but there are a few failures...

On Windows:
```
Test.AggregateTests.Aggregate_AggregateException(labeled: Enumerable.Range, count: 512) [FAIL]
        Assert.Throws() Failure
        Expected: typeof(System.AggregateException)
        Actual:   (No exception was thrown)
        Stack Trace:
           d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.Linq.Parallel\tests\Helpers\Functions.cs(34,0): at Test.Functions.AssertThrowsWrapped[T](Action query)
           d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.Linq.Parallel\tests\QueryOperators\AggregateTests.cs(304,0): at Test.AggregateTests.Aggregate_AggregateException(Labeled`1 labeled, Int32 count)
     Test.AggregateTests.Aggregate_AggregateException(labeled: Partitioner, count: 512) [FAIL]
        Assert.Throws() Failure
        Expected: typeof(System.AggregateException)
        Actual:   (No exception was thrown)
        Stack Trace:
           d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.Linq.Parallel\tests\Helpers\Functions.cs(34,0): at Test.Functions.AssertThrowsWrapped[T](Action query)
           d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.Linq.Parallel\tests\QueryOperators\AggregateTests.cs(300,0): at Test.AggregateTests.Aggregate_AggregateException(Labeled`1 labeled, Int32 count)
```

On Linux:
```
Test.AggregateTests.Aggregate_AggregateException(labeled: Enumerable.Range, count: 512) [FAIL]
02:24:46       Assert.Throws() Failure
02:24:46       Expected: typeof(System.AggregateException)
02:24:46       Actual:   (No exception was thrown)
02:24:46       Stack Trace:
02:24:46             at Test.Functions.AssertThrowsWrapped[T](Action query)
02:24:46             at Test.AggregateTests.Aggregate_AggregateException(Labeled`1 labeled, Int32 count)
```

Clockwork-Muse commented:
Yeah, I reverted all changes on those files.  I can rebase it out if you want?  I thought (because I reverted it) it wouldn't show in the diff/PR.

'Re' because, well I guess it still generates it only once.  Mostly because it changes the returned type.  Should I change it to `BitArrayDelegatEnumerator` or something?
I'm using the RWSlim because I pull a read lock to check the current value, then upgrade if I need to write.  Although since most cases will need to write anyways, I should probably see about a straight lock.

Saw the test failures.  The problem is that the source set isn't always large enough to trigger accumulator combination.  I'm planning on really upping the count (although most stuff dies immediately), but it's something of a Heisenbug.

Side note: adding List and ReadOnlyCollection has added half again to the `OuterLoop` runtime.

stephentoub commented:
> I thought (because I reverted it) it wouldn't show in the diff/PR.

Rather than re-review the entire diff, I was just looking at your recent individual commits.

> I can rebase it out if you want?

We should squash this before merging.  No need to rebase before then.

> Should I change it to BitArrayDelegatEnumerator or something?

It's a nested type for enumerating all of the values in the range, right?  How about just "Enumerator"?

> I'm using the RWSlim because I pull a read lock to check the current value, then upgrade if I need to write. Although since most cases will need to write anyways, I should probably see about a straight lock.

It may not matter, but in my experience, reader-writer locks are only really useful when you have a lot of work to be done concurrently by readers and very few writers.  If there's only a small amount of work to be done by the reader, the extra cost associated with a reader-writer lock often dominates and counteracts the concurrency benefits.

> The problem is that the source set isn't always large enough to trigger accumulator combination. I'm planning on really upping the count

Rather than increasing the count and hoping for the best, could you use a custom partitioner to guarantee that there are multiple partitions and that each has some data?  You should be able to get by with a very small number of elements.   We don't want tests that will sometimes fail based on this or that timing or scheduling issue.

> Side note: adding List and ReadOnlyCollection has added half again to the OuterLoop runtime.

Could we run only a subset of tests with them?

Clockwork-Muse commented:
@stephentoub - been trying to add tests for how many times `Aggregate(...merge...)` is called, but tests for `Enumerable.Range` and `Partitioner` are failing.  Interestingly, `Aggregate(seedFunction ...merge)` _does_ call the seed function DOP-number-of-times, so it seems to want to start that many threads, it just doesn't use them.  It ends up calling the merging a non-deterministic number of times (usually once), although introducing delay into the main aggregation function seems to net more calls.
I dropped the following into `DegreeOfParallelismTests.cs`:

```
[Theory]
[MemberData("DegreeData", new[] { 1, 4 }, new object[] { })]
[MemberData("DegreeData", new[] { 32, 512, 1024, 1024 * 1024 * 4 }, new[] { 4, 16 })]
[OuterLoop]
public static void DegreeOfParallelism_Aggregate_Accumulator(Labeled<ParallelQuery<int>> labeled, int count, int degree)
{
    ParallelQuery<int> query = labeled.Item;
    int accumulatorCombineCount = 0;

    int actual = query.WithDegreeOfParallelism(degree).WithExecutionMode(ParallelExecutionMode.ForceParallelism).Aggregate(
        0,
        (accumulator, x) => accumulator + x,
        (left, right) => { Interlocked.Increment(ref accumulatorCombineCount); return left + right; }, // merge
        result => result);
    Assert.Equal(Functions.SumRange(0, count), actual);
    Assert.Equal(degree - 1, accumulatorCombineCount);
}

[Theory]
[MemberData("DegreeData", new[] { 1, 4 }, new object[] { })]
[MemberData("DegreeData", new[] { 32, 512, 1024, 1024 * 1024 * 4 }, new[] { 4, 16 })]
[OuterLoop]
public static void DegreeOfParallelism_Aggregate_SeedFunction(Labeled<ParallelQuery<int>> labeled, int count, int degree)
{
    ParallelQuery<int> query = labeled.Item;
    int accumulatorCombineCount = 0;
    int seedFunctionCallCount = 0;

    int actual = query.WithDegreeOfParallelism(degree).WithExecutionMode(ParallelExecutionMode.ForceParallelism).Aggregate(
        () => { Interlocked.Increment(ref seedFunctionCallCount); return 0; },
        (accumulator, x) => accumulator + x,
        (left, right) => { Interlocked.Increment(ref accumulatorCombineCount); return left + right; }, // merge
        result => result);
    Assert.Equal(Functions.SumRange(0, count), actual);
    Assert.Equal(seedFunctionCallCount, degree);
    Assert.Equal(seedFunctionCallCount - 1, accumulatorCombineCount);
    Assert.Equal(degree - 1, accumulatorCombineCount);
}
```

... am I just missing something simple I should be doing here?  Should I log a bug?  Or chalk it up to "expected" behavior and check range (`Assert.InRange(accumulatorCombineCount, 0, degree - 1);` for those types?  
Queries sourced from `ParallelEnumerable.Range`, arrays, lists, and `ReadOnlyCollection` (backed by list) are passing this test (turning off load-balancing via `Partitioner.Create(array, false)` also makes partitioners pass, but I don't want to propagate that for all tests).

stephentoub commented:
> am I just missing something simple I should be doing here? Should I log a bug? Or chalk it up to "expected" behaovior

How are you creating the partitioner?  Most likely you're doing it in a way where the results you're seeing are expected.

PLINQ is going to spin up the requested number of workers, each of which is going to enumerate its partition.  The first call to MoveNext for the partition is going to invoke the seed factory to get back the seed value, so the seed factory delegate is invoked before any other elements have been enumerated, and will be invoked even if there are no other elements in the partition.

By default, most of the partitioning schemes employed use load balancing, so just because there are P partitions for N elements doesn't mean each worker is going to process N / P elements.  For many queries, it's likely that the first worker(s) to start running will be able to process most or all of the inputs before the subsequent workers have gotten a chance to start running and processing items, subject to both the runtime's and the OS' scheduling decisions.  This is the primary source of the non-determinism you're seeing.  If a worker goes to enumerate its partition and finds it empty, then it has nothing more to do, but it will have invoked the seed factory.

This is why I suggested using a custom partitioner, one where you could force values to show up in each partition rather than allowing partitions to steal from each other.  Rather than writing your own custom partitioner, you can do it with the Partitioner.Create overload that takes an IList and a loadBalance Boolean.  As you noted, if you set loadBalance to false, your test passes, as each worker should get ~ N / P elements.

Clockwork-Muse commented:
...yeah, that was pretty much what I figured.  My problem though is I've been more interested in calling `AsParallel()` on an enumerable/whatever directly.
I didn't think implementing a custom partitioner wouldn't do me much good unless I also called `WithDegreesOfParallelism(...)`, right?  Because the partitioner is given the number of partitions for it to create (which would have made a lot of unrelated stuff DOP tests).
Guess I'll work up a couple of delegate (non-load-balanced) data-sources for these tests.

stephentoub commented:
> I didn't think implementing a custom partitioner wouldn't do me much good unless I also called WithDegreesOfParallelism(...), right? Because the partitioner is given the number of partitions for it to create (which would have made a lot of unrelated stuff DOP tests

PLINQ always has a fixed DOP, it just defaults to DOP == Environment.ProcessorCount.  WithDegreeOfParallelism simply overrides that to whatever specific value you provide.

Clockwork-Muse commented:
Also had to add a processor-dependent check for Aggregate's AggregateException tests, as I'm not specifying DOP (although I'm pretty sure the build machine has multiple processors).  Doing a wrapping partitioner feels a bit like I'm not testing Enumerable(.Range(...)) itself.  Ah well.

Side note: does Github not have a merge-squash (or rebase-squash) button?  I would have thought _me_ doing it would be a potential security issue... not that I plan on doing anything... +shifty eyes+

stephentoub commented:
>  does Github not have a merge-squash (or rebase-squash) button? I would have thought me doing it would be a potential security issue... not that I plan on doing anything... +shifty eyes+

Heh.  No, unfortunately... I wish it did.  I'll take another pass through after you've squashed.

Clockwork-Muse commented:
>If this approach definitely works for your needs, then ok, but it seems a bit brittle. You're relying on the >underlying Partitioner providing perfect load balancing, such that any element can be stolen by another 
>partition. Consider for example a source partitioner that partitioned an enumerable by having a partition >grab a single element from the source enumerator at a time... with that, your scheme will work fine. But >now consider a partitioner where each partition grabs multiple items from the underlying enumerator at >a time... now with your scheme, it's likely data will be lost. At the very least it'd be worth a comment on >the class as to what it should and should not be used with.
Yeah, I noticed the first time I ran it without specifying the `NoBuffering` option on the underlying partitioner.  At the time I had `StrictPartitioner` written slightly incorrectly, so I had items repeated instead...
Still, will add comment, rebase, squash.

stephentoub commented on src/System.Linq.Parallel/tests/Helpers/IntegerRangeSet.cs:
```diff
@@ -0,0 +1,133 @@
+            lock (_locker)
+            {
+                Assert.False(_seen[entry - _start]);
+                return _seen[entry - _start] = true;
```
What's the point of returning a Boolean here?  Isn't this now just always returning true?

Clockwork-Muse replied:
I use the Boolean in some of the tests, actually.  I use it in `Any(...)` and `All(...)` to verify that it sees all elements (and a few other places too); it's easier to just use the return value than to have to do a multi-statement lambda.

stephentoub commented:
Stephen, I took another skim through it.  Some minor issues, but it generally looks good, and we can fix up other things subsequently as part of another PR... I'd rather just get this in rather than have it remain pending any longer, so I'm going to merge.  Thanks, again, for working to clean these up.  They look much better than they did before.

Clockwork-Muse commented:
@stephentoub - if you pass me a list of outstanding issues, I'll see what I can do about them.

(I'm assuming the custom partitioner is still on the list?)
